File: CS/WebDAVServer.FileSystemStorage.AspNetCore/DavFolder.cs

using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Reflection;
using System.Text.RegularExpressions;
using System.Threading.Tasks;

using ITHit.WebDAV.Server;
using ITHit.WebDAV.Server.Acl;
using ITHit.WebDAV.Server.Class1;
using ITHit.WebDAV.Server.Quota;
using WebDAVServer.FileSystemStorage.AspNetCore.ExtendedAttributes;
using ITHit.WebDAV.Server.Search;
using ITHit.WebDAV.Server.ResumableUpload;
using ITHit.WebDAV.Server.Paging;

namespace WebDAVServer.FileSystemStorage.AspNetCore
{
    /// <summary>
    /// Folder in WebDAV repository.
    /// </summary>
    public class DavFolder : DavHierarchyItem, IFolderAsync, IQuotaAsync, ISearchAsync, IResumableUploadBase
    {

        /// <summary>
        /// Corresponding instance of <see cref="DirectoryInfo"/>.
        /// </summary>
        private readonly DirectoryInfo dirInfo;

        /// <summary>
        /// Returns folder that corresponds to path.
        /// </summary>
        /// <param name="context">WebDAV Context.</param>
        /// <param name="path">Encoded path relative to WebDAV root folder.</param>
        /// <returns>Folder instance or null if physical folder not found in file system.</returns>
        public static async Task<DavFolder> GetFolderAsync(DavContext context, string path)
        {
            string folderPath = context.MapPath(path).TrimEnd(System.IO.Path.DirectorySeparatorChar);
            DirectoryInfo folder = new DirectoryInfo(folderPath);

            // This code blocks vulnerability when "%20" folder can be injected into path and folder.Exists returns 'true'.
            if (!folder.Exists || string.Compare(folder.FullName.TrimEnd(System.IO.Path.DirectorySeparatorChar), folderPath, StringComparison.OrdinalIgnoreCase) != 0)
            {
                return null;
            }

            return new DavFolder(folder, context, path);
        }

        /// <summary>
        /// Initializes a new instance of this class.
        /// </summary>
        /// <param name="directory">Corresponding folder in the file system.</param>
        /// <param name="context">WebDAV Context.</param>
        /// <param name="path">Encoded path relative to WebDAV root folder.</param>
        protected DavFolder(DirectoryInfo directory, DavContext context, string path)
            : base(directory, context, path.TrimEnd('/') + "/")
        {
            dirInfo = directory;
        }   

        /// <summary>
        /// Called when children of this folder with paging information are being listed.
        /// </summary>
        /// <param name="propNames">List of properties to retrieve with the children. They will be queried by the engine later.</param>
        /// <param name="offset">The number of children to skip before returning the remaining items. Start listing from from next item.</param>
        /// <param name="nResults">The number of items to return.</param>
        /// <param name="orderProps">List of order properties requested by the client.</param>
        /// <returns>Items requested by the client and a total number of items in this folder.</returns>
        public virtual async Task<PageResults> GetChildrenAsync(IList<PropertyName> propNames, long? offset, long? nResults, IList<OrderProperty> orderProps)
        {
            // Enumerates all child files and folders.
            // You can filter children items in this implementation and 
            // return only items that you want to be visible for this 
            // particular user.

            IList<IHierarchyItemAsync> children = new List<IHierarchyItemAsync>();

            FileSystemInfo[] fileInfos = null;
            long totalItems = 0;
            fileInfos = dirInfo.GetFileSystemInfos();
            totalItems = fileInfos.Length;

            // Apply sorting.
            fileInfos = SortChildren(fileInfos, orderProps);

            // Apply paging.
            if (offset.HasValue && nResults.HasValue)
            {
                fileInfos = fileInfos.Skip((int)offset.Value).Take((int)nResults.Value).ToArray();
            }

            foreach (FileSystemInfo fileInfo in fileInfos)
            {
                string childPath = Path + EncodeUtil.EncodeUrlPart(fileInfo.Name);
                IHierarchyItemAsync child = await context.GetHierarchyItemAsync(childPath);
                if (child != null)
                {
                    children.Add(child);
                }
            }

            return new PageResults(children, totalItems);
        }

        /// <summary>
        /// Called when a new file is being created in this folder.
        /// </summary>
        /// <param name="name">Name of the new file.</param>
        /// <returns>The new file.</returns>
        public async Task<IFileAsync> CreateFileAsync(string name)
        {
            await RequireHasTokenAsync();
            string fileName = System.IO.Path.Combine(fileSystemInfo.FullName, name);

            using (FileStream stream = new FileStream(fileName, FileMode.CreateNew))
            {
            }
            await context.socketService.NotifyRefreshAsync(Path);

            return (IFileAsync)await context.GetHierarchyItemAsync(Path + EncodeUtil.EncodeUrlPart(name));
        }

        /// <summary>
        /// Called when a new folder is being created in this folder.
        /// </summary>
        /// <param name="name">Name of the new folder.</param>
        virtual public async Task CreateFolderAsync(string name)
        {
            await RequireHasTokenAsync();
            dirInfo.CreateSubdirectory(name);
            await context.socketService.NotifyRefreshAsync(Path);
        }

        /// <summary>
        /// Called when this folder is being copied.
        /// </summary>
        /// <param name="destFolder">Destination parent folder.</param>
        /// <param name="destName">New folder name.</param>
        /// <param name="deep">Whether children items shall be copied.</param>
        /// <param name="multistatus">Information about child items that failed to copy.</param>
        public override async Task CopyToAsync(IItemCollectionAsync destFolder, string destName, bool deep, MultistatusException multistatus)
        {
            DavFolder targetFolder = destFolder as DavFolder;
            if (targetFolder == null)
            {
                throw new DavException("Target folder doesn't exist", DavStatus.CONFLICT);
            }

            if (IsRecursive(targetFolder))
            {
                throw new DavException("Cannot copy to subfolder", DavStatus.FORBIDDEN);
            }

            string newDirLocalPath = System.IO.Path.Combine(targetFolder.FullPath, destName);
            string targetPath = targetFolder.Path + EncodeUtil.EncodeUrlPart(destName);

            // Create folder at the destination.
            try
            {
                if (!Directory.Exists(newDirLocalPath))
                {
                    await targetFolder.CreateFolderAsync(destName);
                }
            }
            catch (DavException ex)
            {
                // Continue, but report error to client for the target item.
                multistatus.AddInnerException(targetPath, ex);
            }

            // Copy children.
            IFolderAsync createdFolder = (IFolderAsync)await context.GetHierarchyItemAsync(targetPath);
            foreach (DavHierarchyItem item in (await GetChildrenAsync(new PropertyName[0], null, null, null)).Page)
            {
                if (!deep && item is DavFolder)
                {
                    continue;
                }

                try
                {
                    await item.CopyToAsync(createdFolder, item.Name, deep, multistatus);
                }
                catch (DavException ex)
                {
                    // If a child item failed to copy we continue but report error to client.
                    multistatus.AddInnerException(item.Path, ex);
                }
            }
            await context.socketService.NotifyRefreshAsync(targetFolder.Path);
        }

        /// <summary>
        /// Called when this folder is being moved or renamed.
        /// </summary>
        /// <param name="destFolder">Destination folder.</param>
        /// <param name="destName">New name of this folder.</param>
        /// <param name="multistatus">Information about child items that failed to move.</param>
        public override async Task MoveToAsync(IItemCollectionAsync destFolder, string destName, MultistatusException multistatus)
        {
            await RequireHasTokenAsync();
            DavFolder targetFolder = destFolder as DavFolder;
            if (targetFolder == null)
            {
                throw new DavException("Target folder doesn't exist", DavStatus.CONFLICT);
            }

            if (IsRecursive(targetFolder))
            {
                throw new DavException("Cannot move folder to its subtree.", DavStatus.FORBIDDEN);
            }

            string newDirPath = System.IO.Path.Combine(targetFolder.FullPath, destName);
            string targetPath = targetFolder.Path + EncodeUtil.EncodeUrlPart(destName);

            try
            {
                // Remove item with the same name at destination if it exists.
                IHierarchyItemAsync item = await context.GetHierarchyItemAsync(targetPath);
                if (item != null)
                    await item.DeleteAsync(multistatus);

                await targetFolder.CreateFolderAsync(destName);
            }
            catch (DavException ex)
            {
                // Continue the operation but report error with destination path to client.
                multistatus.AddInnerException(targetPath, ex);
                return;
            }

            // Move child items.
            bool movedSuccessfully = true;
            IFolderAsync createdFolder = (IFolderAsync)await context.GetHierarchyItemAsync(targetPath);
            foreach (DavHierarchyItem item in (await GetChildrenAsync(new PropertyName[0], null, null, null)).Page)
            {
                try
                {
                    await item.MoveToAsync(createdFolder, item.Name, multistatus);
                }
                catch (DavException ex)
                {
                    // Continue the operation but report error with child item to client.
                    multistatus.AddInnerException(item.Path, ex);
                    movedSuccessfully = false;
                }
            }

            if (movedSuccessfully)
            {
                await DeleteAsync(multistatus);
            }
            // Refresh client UI.
            await context.socketService.NotifyDeleteAsync(Path);
            await context.socketService.NotifyRefreshAsync(GetParentPath(targetPath));
        }

        /// <summary>
        /// Called whan this folder is being deleted.
        /// </summary>
        /// <param name="multistatus">Information about items that failed to delete.</param>
        public override async Task DeleteAsync(MultistatusException multistatus)
        {
            /*
            if (await GetParentAsync() == null)
            {
                throw new DavException("Cannot delete root.", DavStatus.NOT_ALLOWED);
            }
            */
            await RequireHasTokenAsync();
            bool allChildrenDeleted = true;
            foreach (IHierarchyItemAsync child in (await GetChildrenAsync(new PropertyName[0], null, null, null)).Page)
            {
                try
                {
                    await child.DeleteAsync(multistatus);
                }
                catch (DavException ex)
                {
                    //continue the operation if a child failed to delete. Tell client about it by adding to multistatus.
                    multistatus.AddInnerException(child.Path, ex);
                    allChildrenDeleted = false;
                }
            }

            if (allChildrenDeleted)
            {
                dirInfo.Delete();
                await context.socketService.NotifyDeleteAsync(Path);
            }
        }

        /// <summary>
        /// Returns free bytes available to current user.
        /// </summary>
        /// <returns>Free bytes available.</returns>
        public async Task<long> GetAvailableBytesAsync()
        {
            // Here you can return amount of bytes available for current user.
            // For the sake of simplicity we return entire available disk space.

            // Note: NTFS quotes retrieval for current user works very slowly.

            return await dirInfo.GetStorageFreeBytesAsync();
        }

        /// <summary>
        /// Returns used bytes by current user.
        /// </summary>
        /// <returns>Number of bytes used on disk.</returns>
        public async Task<long> GetUsedBytesAsync()
        {
            // Here you can return amount of bytes used by current user.
            // For the sake of simplicity we return entire used disk space.

            //Note: NTFS quotes retrieval for current user works very slowly.

            return await dirInfo.GetStorageUsedBytesAsync();
        }

        /// <summary>
        /// Searches files and folders in current folder using search phrase, offset, nResults and options.
        /// </summary>
        /// <param name="searchString">A phrase to search.</param>
        /// <param name="options">Search options.</param>
        /// <param name="propNames">
        /// List of properties to retrieve with each item returned by this method. They will be requested by the 
        /// Engine in <see cref="IHierarchyItemAsync.GetPropertiesAsync(IList{PropertyName}, bool)"/> call.
        /// </param>
        /// <param name="offset">The number of children to skip before returning the remaining items. Start listing from from next item.</param>
        /// <param name="nResults">The number of items to return.</param>
        /// <returns>List of <see cref="IHierarchyItemAsync"/> satisfying search request.</returns>1
        /// <returns>Items satisfying search request and a total number.</returns>
        public async Task<PageResults> SearchAsync(string searchString, SearchOptions options, List<PropertyName> propNames, long? offset, long? nResults)
        {
            // Not implemented currently. .NET Core on Mac OS X does not provide a OLEDB driver at this time.
            // To generate a Windows-specific search code use the 'ASP.NET WebDAV Server Application' wizard for Visual Studio.
            return new PageResults(null, 0);
        }

        /// <summary>
        /// Determines whether <paramref name="destFolder"/> is inside this folder.
        /// </summary>
        /// <param name="destFolder">Folder to check.</param>
        /// <returns>Returns <c>true</c> if <paramref name="destFolder"/> is inside thid folder.</returns>
        private bool IsRecursive(DavFolder destFolder)
        {
            return destFolder.Path.StartsWith(Path);
        }

        /// <summary>
        /// Sorts array of FileSystemInfo according to the specified order.
        /// </summary>
        /// <param name="fileInfos">Array of files and folders to sort.</param>
        /// <param name="orderProps">Sorting order.</param>
        /// <returns>Sorted list of files and folders.</returns>
        private FileSystemInfo[] SortChildren(FileSystemInfo[] fileInfos, IList<OrderProperty> orderProps)
        {
            if (orderProps != null && orderProps.Count() != 0)
            {
                // map DAV properties to FileSystemInfo 
                Dictionary<string, string> mappedProperties = new Dictionary<string, string>()
                { { "displayname", "Name" }, { "getlastmodified", "LastWriteTime" }, { "getcontenttype", "Extension" },
                  { "quota-used-bytes", "ContentLength" }, { "is-directory", "IsDirectory" } };
                IOrderedEnumerable<FileSystemInfo> orderedFileInfos = null;
                int index = 0;

                foreach (OrderProperty ordProp in orderProps)
                {
                    string propertyName = mappedProperties[ordProp.Property.Name];
                    Func<FileSystemInfo, object> sortFunc = null;
                    PropertyInfo propertyInfo = (typeof(FileSystemInfo)).GetProperties().FirstOrDefault(p => p.Name.Equals(propertyName, StringComparison.InvariantCultureIgnoreCase));
                    if (propertyInfo != null)
                    {
                        sortFunc = p => p.GetType().GetProperty(propertyInfo.Name).GetValue(p);
                    }
                    else if (propertyName == "IsDirectory")
                    {
                        sortFunc = p => p.IsDirectory();
                    }
                    else if (propertyName == "ContentLength")
                    {
                        sortFunc = p => p is FileInfo ? (p as FileInfo).Length : 0;
                    }

                    if (sortFunc != null)
                    {
                        if (index++ == 0)
                        {
                            if (ordProp.Ascending)
                                orderedFileInfos = fileInfos.OrderBy(sortFunc);
                            else
                                orderedFileInfos = fileInfos.OrderByDescending(sortFunc);
                        }
                        else
                        {
                            if (ordProp.Ascending)
                                orderedFileInfos = orderedFileInfos.ThenBy(sortFunc);
                            else
                                orderedFileInfos = orderedFileInfos.ThenByDescending(sortFunc);
                        }
                    }

                }

                if (orderedFileInfos != null)
                {
                    fileInfos = orderedFileInfos.ToArray();
                }
            }

            return fileInfos;
        }
    }
}

File: VB/WebDAVServer.SqlStorage.AspNet/DavFolder.vb
Imports System
Imports System.Collections.Generic
Imports System.IO
Imports System.Linq
Imports System.Threading.Tasks
Imports ITHit.WebDAV.Server
Imports ITHit.WebDAV.Server.Class1
Imports ITHit.WebDAV.Server.Class2
Imports ITHit.WebDAV.Server.Quota
Imports ITHit.WebDAV.Server.Search

''' <summary>
''' Represents folder in webdav repository.
''' </summary>
Public Class DavFolder
    Inherits DavHierarchyItem
    Implements IFolderAsync, IQuotaAsync, ISearchAsync

    ''' <summary>
    ''' Initializes a new instance of the <see cref="DavFolder"/>  class.
    ''' </summary>
    ''' <param name="context">Instance of <see cref="DavContext"/>  class.</param>
    ''' <param name="itemId">Id of this folder.</param>
    ''' <param name="parentId">Id of parent folder.</param>
    ''' <param name="name">Name of this folder.</param>
    ''' <param name="path">Encoded WebDAV path to this folder.</param>
    ''' <param name="created">Date when the folder was created.</param>
    ''' <param name="modified">Date when the folder was modified.</param>
    ''' <param name="fileAttributes">File attributes of the folder (hidden, read-only etc.)</param>
    Public Sub New(context As DavContext,
                  itemId As Guid,
                  parentId As Guid,
                  name As String,
                  path As String,
                  created As DateTime,
                  modified As DateTime, fileAttributes As FileAttributes)
        MyBase.New(context, itemId, parentId, name, path, created, modified, fileAttributes)
    End Sub

    ''' <summary>
    ''' Gets child items of this folder (files or folders).
    ''' </summary>
    ''' <param name="props">
    ''' List of properties which will be requested from children later. We don't use it here.
    ''' </param>
    ''' <returns>Enumerable with files and folders.</returns>
    Public Async Function GetChildrenAsync(props As IList(Of PropertyName)) As Task(Of IEnumerable(Of IHierarchyItemAsync)) Implements IItemCollectionAsync.GetChildrenAsync
        Dim command As String = "SELECT 
                      ItemId
                    , ParentItemId
                    , ItemType
                    , Name
                    , Created
                    , Modified                    , FileAttributes                   FROM Item
                   WHERE ParentItemId = @Parent"
        Return Await Context.ExecuteItemAsync(Of IHierarchyItemAsync)(Path,
                                                                     command,
                                                                     "@Parent", ItemId)
    End Function

    ''' <summary>
    ''' Creates file with specified name in this folder.
    ''' </summary>
    ''' <param name="name">File name.</param>
    ''' <returns>Instance of <see cref="File"/>  referring to newly created file.</returns>
    Public Async Function CreateFileAsync(name As String) As Task(Of IFileAsync) Implements IFolderAsync.CreateFileAsync
        If Not Await ClientHasTokenAsync() Then
            Throw New LockedException()
        End If

        Dim child = Await createChildAsync(name, ItemType.File)
        Await Context.socketService.NotifyRefreshAsync(Path)
        Return CType(child, IFileAsync)
    End Function

    ''' <summary>
    ''' Creates folder with specified name in this folder.
    ''' </summary>
    ''' <param name="name">Name of folder to be created.</param>
    Public Async Function CreateFolderAsync(name As String) As Task Implements IFolderAsync.CreateFolderAsync
        If Not Await ClientHasTokenAsync() Then
            Throw New LockedException()
        End If

        Await createChildAsync(name, ItemType.Folder)
        Await Context.socketService.NotifyRefreshAsync(Path)
    End Function

    ''' <summary>
    ''' Copies this folder to another folder with option to rename it.
    ''' </summary>
    ''' <param name="destFolder">Folder to copy this folder to.</param>
    ''' <param name="destName">New name of this folder.</param>
    ''' <param name="deep">Whether children shall be copied.</param>
    ''' <param name="multistatus">Container for errors. We put here errors which occur with
    ''' individual items being copied.</param>
    Public Overrides Async Function CopyToAsync(destFolder As IItemCollectionAsync,
                                               destName As String,
                                               deep As Boolean,
                                               multistatus As MultistatusException) As Task Implements IHierarchyItemAsync.CopyToAsync
        Dim destDavFolder As DavFolder = TryCast(destFolder, DavFolder)
        If destFolder Is Nothing Then
            Throw New DavException("Destination folder doesn't exist", DavStatus.CONFLICT)
        End If

        If Not Await destDavFolder.ClientHasTokenAsync() Then
            Throw New LockedException("Doesn't have token for destination folder.")
        End If

        If isRecursive(destDavFolder) Then
            Throw New DavException("Cannot copy folder to its subtree", DavStatus.FORBIDDEN)
        End If

        Dim destItem As IHierarchyItemAsync = Await destDavFolder.FindChildAsync(destName)
        If destItem IsNot Nothing Then
            Try
                Await destItem.DeleteAsync(multistatus)
            Catch ex As DavException
                multistatus.AddInnerException(destItem.Path, ex)
                Return
            End Try
        End If

        Dim newDestFolder As DavFolder = Await CopyThisItemAsync(destDavFolder, Nothing, destName)
        ' copy children
        If deep Then
            For Each child As IHierarchyItemAsync In Await GetChildrenAsync(New PropertyName(-1) {})
                Dim dbchild = TryCast(child, DavHierarchyItem)
                Try
                    Await dbchild.CopyToAsync(newDestFolder, child.Name, deep, multistatus)
                Catch ex As DavException
                    multistatus.AddInnerException(dbchild.Path, ex)
                End Try
            Next
        End If

        Await Context.socketService.NotifyRefreshAsync(newDestFolder.Path)
    End Function

    ''' <summary>
    ''' Moves this folder to destination folder with option to rename.
    ''' </summary>
    ''' <param name="destFolder">Folder to copy this folder to.</param>
    ''' <param name="destName">New name of this folder.</param>
    ''' <param name="multistatus">Container for errors. We put here errors occurring while moving
    ''' individual files/folders.</param>
    Public Overrides Async Function MoveToAsync(destFolder As IItemCollectionAsync, destName As String, multistatus As MultistatusException) As Task Implements IHierarchyItemAsync.MoveToAsync
        Dim destDavFolder As DavFolder = TryCast(destFolder, DavFolder)
        If destFolder Is Nothing Then
            Throw New DavException("Destination folder doesn't exist", DavStatus.CONFLICT)
        End If

        If isRecursive(destDavFolder) Then
            Throw New DavException("Cannot move folder to its subtree", DavStatus.FORBIDDEN)
        End If

        Dim parent As DavFolder = Await GetParentAsync()
        If parent Is Nothing Then
            Throw New DavException("Cannot move root", DavStatus.CONFLICT)
        End If

        If Not Await ClientHasTokenAsync() OrElse Not Await destDavFolder.ClientHasTokenAsync() OrElse Not Await parent.ClientHasTokenAsync() Then
            Throw New LockedException()
        End If

        Dim destItem As DavHierarchyItem = Await destDavFolder.FindChildAsync(destName)
        Dim newDestFolder As DavFolder
        ' copy this folder
        If destItem IsNot Nothing Then
            If TypeOf destItem Is IFileAsync Then
                Try
                    Await destItem.DeleteAsync(multistatus)
                Catch ex As DavException
                    multistatus.AddInnerException(destItem.Path, ex)
                    Return
                End Try

                newDestFolder = Await CopyThisItemAsync(destDavFolder, Nothing, destName)
            Else
                newDestFolder = TryCast(destItem, DavFolder)
                If newDestFolder Is Nothing Then
                    multistatus.AddInnerException(destItem.Path,
                                                 New DavException("Destionation item is not folder", DavStatus.CONFLICT))
                End If
            End If
        Else
            newDestFolder = Await CopyThisItemAsync(destDavFolder, Nothing, destName)
        End If

        ' move children
        Dim movedAllChildren As Boolean = True
        For Each child As IHierarchyItemAsync In Await GetChildrenAsync(New PropertyName(-1) {})
            Dim dbchild As DavHierarchyItem = TryCast(child, DavHierarchyItem)
            Try
                Await dbchild.MoveToAsync(newDestFolder, child.Name, multistatus)
            Catch ex As DavException
                multistatus.AddInnerException(dbchild.Path, ex)
                movedAllChildren = False
            End Try
        Next

        If movedAllChildren Then
            Await DeleteThisItemAsync(parent)
        End If

        ' Refresh client UI.
        Await Context.socketService.NotifyDeleteAsync(Path)
        Await Context.socketService.NotifyRefreshAsync(GetParentPath(newDestFolder.Path))
    End Function

    ''' <summary>
    ''' Deletes this folder.
    ''' </summary>
    ''' <param name="multistatus">Container for errors.
    ''' If some child file/folder fails to remove we report error in this container.</param>
    Public Overrides Async Function DeleteAsync(multistatus As MultistatusException) As Task Implements IHierarchyItemAsync.DeleteAsync
        Dim parent As DavFolder = Await GetParentAsync()
        If parent Is Nothing Then
            Throw New DavException("Cannot delete root.", DavStatus.CONFLICT)
        End If

        If Not Await parent.ClientHasTokenAsync() Then
            Throw New LockedException()
        End If

        If Not Await ClientHasTokenAsync() Then
            Throw New LockedException()
        End If

        Dim deletedAllChildren As Boolean = True
        For Each child As IHierarchyItemAsync In Await GetChildrenAsync(New PropertyName(-1) {})
            Dim dbchild As DavHierarchyItem = TryCast(child, DavHierarchyItem)
            Try
                Await dbchild.DeleteAsync(multistatus)
            Catch ex As DavException
                multistatus.AddInnerException(dbchild.Path, ex)
                deletedAllChildren = False
            End Try
        Next

        If deletedAllChildren Then
            Await DeleteThisItemAsync(parent)
            Await Context.socketService.NotifyDeleteAsync(Path)
        End If
    End Function

    ''' <summary>
    ''' Returns free bytes available to current user.
    ''' </summary>
    ''' <returns>Free bytes available.</returns>
    Public Async Function GetAvailableBytesAsync() As Task(Of Long) Implements IQuotaAsync.GetAvailableBytesAsync
        'let's assume total space is 5GB.
        Return(5L * 1024 * 1024 * 1024) - Await GetUsedBytesAsync()
    End Function

    ''' <summary>
    ''' Returns used bytes by current user.
    ''' </summary>
    ''' <returns>Number of bytes consumed by files of current user.</returns>
    Public Async Function GetUsedBytesAsync() As Task(Of Long) Implements IQuotaAsync.GetUsedBytesAsync
        Return Await Context.ExecuteScalarAsync(Of Long)("SELECT SUM(DATALENGTH(Content)) FROM Item")
    End Function

    ''' <summary>
    ''' Searches files and folders in current folder using search phrase and options.
    ''' </summary>
    ''' <param name="searchString">A phrase to search.</param>
    ''' <param name="options">Search options.</param>
    ''' <param name="propNames">
    ''' List of properties to retrieve with each item returned by this method. They will be requested by the 
    ''' Engine in <see cref="IHierarchyItemAsync.GetPropertiesAsync(IList{PropertyName}, bool)"/>  call.
    ''' </param>
    ''' <returns>List of <see cref="IHierarchyItemAsync"/>  satisfying search request.</returns>
    Public Async Function SearchAsync(searchString As String, options As SearchOptions, propNames As List(Of PropertyName)) As Task(Of IEnumerable(Of IHierarchyItemAsync)) Implements ISearchAsync.SearchAsync
        Dim includeSnippet As Boolean = propNames.Any(Function(s) s.Name = SNIPPET)
        Dim condition As String = "Name LIKE @Name"
        ' To enable full-text search, uncoment the code below and follow instructions 
        ' in DB.sql to enable full-text indexing
        Dim commandText As String = [String].Format("SELECT 
                      ItemId
                    , ParentItemId
                    , ItemType
                    , Name
                    , Created
                    , Modified
                    , FileAttributes                          
                   FROM Item
                   WHERE ParentItemId = @Parent AND ({0})", condition)
        Dim result As List(Of IHierarchyItemAsync) = New List(Of IHierarchyItemAsync)()
        Await GetSearchResultsAsync(result, commandText, searchString, includeSnippet)
        Return result
    End Function

    ''' <summary>
    ''' Produces recursive search in current folder.
    ''' </summary>
    ''' <param name="result">A list to add search results to.</param>
    ''' <param name="commandText">SQL command text for search in a folder.</param>
    ''' <param name="searchString">A phrase to search.</param>
    Private Async Function GetSearchResultsAsync(result As List(Of IHierarchyItemAsync), commandText As String, searchString As String, includeSnippet As Boolean) As Task
        ' search this folder
        Dim folderSearchResults As IEnumerable(Of IHierarchyItemAsync) = Await Context.ExecuteItemAsync(Of IHierarchyItemAsync)(Path,
                                                                                                                               commandText,
                                                                                                                               "@Parent", ItemId,
                                                                                                                               "@Name", searchString,
                                                                                                                               "@Content", searchString)
        For Each item As IHierarchyItemAsync In folderSearchResults
            If includeSnippet AndAlso TypeOf item Is DavFile Then TryCast(item, DavFile).Snippet = "Not Implemented"
        Next

        result.AddRange(folderSearchResults)
        ' search children
        For Each item As IHierarchyItemAsync In Await GetChildrenFoldersAsync()
            Dim folder As DavFolder = TryCast(item, DavFolder)
            If folder IsNot Nothing Then Await folder.GetSearchResultsAsync(result, commandText, searchString, includeSnippet)
        Next
    End Function

    ''' <summary>
    ''' Gets the children of current folder (non-recursive).
    ''' </summary>
    ''' <returns>The children folders of current folder.</returns>
    Public Async Function GetChildrenFoldersAsync() As Task(Of IEnumerable(Of IHierarchyItemAsync))
        Dim command As String = "SELECT 
                      ItemId
                    , ParentItemId
                    , ItemType
                    , Name
                    , Created
                    , Modified
                    , FileAttributes
                   FROM Item
                   WHERE ParentItemId = @Parent AND ItemType = 3"
        Return Await Context.ExecuteItemAsync(Of IHierarchyItemAsync)(Path,
                                                                     command,
                                                                     "@Parent", ItemId)
    End Function

    ''' <summary>
    ''' Finds file or folder with specified name inside this folder.
    ''' </summary>
    ''' <param name="childName">Name of child to find.</param>
    ''' <returns>Instance of <see cref="DavHierarchyItem"/>  or <c>null</c>.</returns>
    Friend Async Function FindChildAsync(childName As String) As Task(Of DavHierarchyItem)
        Dim commandText As String = "SELECT
                     ItemId
                   , ParentItemId
                   , ItemType
                   , Name
                   , Created
                   , Modified 
                   , FileAttributes  
                  FROM Item
                  WHERE ParentItemId = @Parent
                  AND Name = @Name"
        Dim davHierarchyItems As IList(Of DavHierarchyItem) = Await Context.ExecuteItemAsync(Of DavHierarchyItem)(Path,
                                                                                                                 commandText,
                                                                                                                 "@Parent", ItemId,
                                                                                                                 "@Name", childName)
        Return davHierarchyItems.FirstOrDefault()
    End Function

    ''' <summary>
    ''' Determines whether the client has submitted lock tokens for all locked files in the subtree.
    ''' </summary>
    ''' <returns>Returns <c>true</c> if lock tockens for all locked files in the subtree are submitted.</returns>
    Friend Async Function ClientHasTokenForTreeAsync() As Task(Of Boolean)
        If Not Await ClientHasTokenAsync() Then
            Return False
        End If

        For Each child As IHierarchyItemAsync In Await GetChildrenAsync(New PropertyName(-1) {})
            Dim childFolder As DavFolder = TryCast(child, DavFolder)
            If childFolder IsNot Nothing Then
                If Not Await childFolder.ClientHasTokenForTreeAsync() Then
                    Return False
                End If
            Else
                Dim childItem As DavHierarchyItem = TryCast(child, DavHierarchyItem)
                If Not Await childItem.ClientHasTokenAsync() Then
                    Return False
                End If
            End If
        Next

        Return True
    End Function

    ''' <summary>
    ''' Determines whether <paramref name="destFolder"/>  is inside this folder.
    ''' </summary>
    ''' <param name="destFolder">Folder to test.</param>
    ''' <returns>Returns <c>true</c>if <paramref name="destFolder"/>  is inside this folder.</returns>
    Private Function isRecursive(destFolder As DavFolder) As Boolean
        Return destFolder.Path.StartsWith(Path)
    End Function

    ''' <summary>
    ''' Creates file or folder with specified name inside this folder.
    ''' </summary>
    ''' <param name="name">File/folder name.</param>
    ''' <param name="itemType">Type of item: file or folder.</param>
    ''' <returns>Newly created item.</returns>
    Private Async Function createChildAsync(name As String, itemType As ItemType) As Task(Of DavHierarchyItem)
        Dim newID As Guid = Guid.NewGuid()
        Dim commandText As String = "INSERT INTO Item(
                      ItemId
                    , Name
                    , Created
                    , Modified
                    , ParentItemId
                    , ItemType
                    , TotalContentLength
                    , FileAttributes
                    )
                VALUES(
                     @Identity
                   , @Name
                   , GETUTCDATE()
                   , GETUTCDATE()
                   , @Parent
                   , @ItemType
                   , 0
                   , @FileAttributes
                   )"
        Await Context.ExecuteNonQueryAsync(commandText,
                                          "@Name", name,
                                          "@Parent", ItemId,
                                          "@ItemType", itemType,
                                          "@FileAttributes", (If(itemType = ItemType.Folder, CInt(FileAttributes.Directory), CInt(FileAttributes.Normal))),
                                          "@Identity", newID)
        'UpdateModified(); do not update time for folder as transaction will block concurrent files upload
        Select Case itemType
            Case ItemType.File
                Return New DavFile(Context,
                                  newID,
                                  ItemId,
                                  name,
                                  Path & EncodeUtil.EncodeUrlPart(name),
                                  DateTime.UtcNow,
                                  DateTime.UtcNow, FileAttributes.Normal)
            Case ItemType.Folder
                ' do not need to return created folder
                Return Nothing
            Case Else
                Return Nothing
        End Select
    End Function
End Class
